### PR TITLE
Implicit exit/entrance conditions and related docs

### PR DESCRIPTION
**Pull request by blkerby**
All these implicit strats are currently implemented in Map Rando with the exception of the implicit `comeInWithMockball`. So, with that exception, this PR is an attempt to document the existing way that we are interpreting things rather than add new implicit strats. It also adds a way to override these implicit strats, which should allow us to eliminate the need for `spawnAt` nodes in many rooms, and eventually perhaps eliminate the `spawnAt` property altogether.

Something that will still need refining later is how to address the fact that it is unsound for the logic to assume you can enter a heated room (with an implicit "comeInNormally") and leave (with an implicit "leaveNormally") with no heat damage. This is a problem that exists in the randomizer now, though we've never seen an unbeatable seed occur because of it. We could add 40 heat frames to the implicit "leaveNormally" to solve the unsoundness in this case, but then it would be adding an unnecessary heat frame requirement to other strats that approach the door from a different location (not entering from the same door). In general, in heated rooms there is some ambiguity about what exactly it means to be at a door node, which should be clarified at some point, though the way to do this is not yet obvious.

### CODE REVIEW

File: strats.md [hunk 1]
```diff
@@ -1516,6 +1498,152 @@ The `endsWithShineCharge` property indicates that a strat ends while in a shinec
...
+    }
+  }
+}
+```
+
+As with other `comeInWithGMode` strats having `"morphed": true`, it is assumed here that Samus morphs by either using the Morph item or by obtaining artificial morph. Also note that if there are strats from this door node to itself with `"gModeRegainMobility": true`, then using those strats to regain mobility are considered to be an option as part of this implicit strat, again the same as with other strats having a `comeInWithGMode` entrance condition.
+
+This implicit strat can be disabled by setting the node property `"useImplicitCarryGModeMorphBackThrough": false`.
+
+If the node has a `spawnAt` property, or if it has `"isDoorImmediatelyClosed": true`, or if the node is a vertical door in bottom position (leading up), then this implicit strat is disabled unless the node is explicitly marked as `"useImplicitCarryGModeMorphBackThrough": true`. Note that the exception about the node being a door in bottom position differs from unmorphed implicit strats for carrying G-mode back a door, described in the previous section, which have no such exception.
```
Comment (kjbranch) [inline strats.md:1645]:
```suggestion
If the node has a `spawnAt` property, or if it has `"isDoorImmediatelyClosed": true`, or if the node is a vertical door in bottom position (leading up), then this implicit strat is disabled unless the node is explicitly marked as `"useImplicitCarryGModeMorphBackThrough": true`. Note that the exception about the node being a door in bottom position differs from unmorphed implicit strats for carrying G-mode back through a door, described in the previous section, which have no such exception.
```

**Review by kjbranch — APPROVED** [review]
This looks fine, but I do wonder if we should just remove spawnAt's. I dont think they really serve a purpose after this migration.

File: strats.md [hunk 2]
```diff
@@ -1516,6 +1498,152 @@ The `endsWithShineCharge` property indicates that a strat ends while in a shinec
 
 This property should not be combined with a `leaveShinecharged` exit condition.
 
+## Implicit Entrance and Exit Strats
+
+By default, door nodes come with implicit strats for entering and exiting the room through the door. These implicit strats can individually be disabled by setting certain node properties, as described below in detail for each kind of implicit strat. Disabling an implicit strat can make sense if enemies or other hazards are near the door, making the door not free to use.
+
+In this section, a door node refers to a node with `"nodeType": "door"`. Likewise, an entrance node refers to a node with `"nodeType": "entrance"`, and an exit node refers to a node with `"nodeType": "exit"`.
```
[the remark this answers is not in the document]
Comment (blkerby) [reply strats.md:1505]: Just checked, and looks like elevators also have `"nodeType": "door"`, they're just distinguished by `"nodeSubType": "elevator"`. I wasn't thinking about elevators when I made the PR but the rules described seem to still make sense for them in general, so I think it's ok? Either way, I think the heated elevators are going to need to be overriding the implicit entrance/exit strats.

File: strats.md [hunk 3]
```diff
@@ -1516,6 +1498,152 @@ The `endsWithShineCharge` property indicates that a strat ends while in a shinec
...
+    {"heatFrames": 10}
+  ]
+}
+```
+
+This implicit strat is to cover cases where leaving with a mockball may be desirable or unavoidable in the previous room, but where the mockball may be unhelpful in the current room. The 10 heat frames are to cover the time potentially needed to stop and unmorph after entering the room.
+
+This implicit strat can be disabled by setting the node property `"useImplicitComeInWithMockball": false`.
+
+If the node has a `spawnAt` property, then this implicit strat is disabled unless the node is explicitly marked as `"useImplicitComeInWithMockball": true`.
```
[the remark this answers is not in the document]
Comment (blkerby) [reply strats.md]: Good point. Yeah this kind of special case is one of the reasons I want to get rid of spawnAt, so we don't have to answer questions like this. For now, I just reworded the doc to say that the implicit strat shouldn't be enabled if there is a spawnAt.

**Comment by blkerby** [comment]
> This looks fine, but I do wonder if we should just remove spawnAt's. I dont think they really serve a purpose after this migration.

I think removing `spawnAt` sounds good but figured it should probably be looked at room-by-room in a separate PR (or possibly multiple PRs). Assuming they all get removed, then they could be deleted from the schema & docs at that point, if that sounds like a good plan?

**Review by osse101 — APPROVED** [review]

**Pull request merged**